File: sdk/python/combee/cells.py
"""Cells / SQL / KV / Backups / Replication(同步 + 异步等价)。"""

from __future__ import annotations

from typing import Any, Optional


class Sql:
    def __init__(self, http, cell_id: str):
        self._http = http
        self._cell = cell_id

    def _path(self) -> str:
        return f"/v1/databases/{self._cell}/sql"

    async def _apath(self) -> str:
        return self._path()

    def _map_rows(self, raw: dict) -> dict:
        columns = raw.get("columns", [])
        rows = [
            dict(zip(columns, row)) if isinstance(row, list) else row
            for row in raw.get("rows", [])
        ]
        return {"columns": columns, "rows": rows}

    @staticmethod
    def _body(sql: str, params: Optional[list]) -> dict:
        # 服务端不接受 params: null;无参数时省略字段
        body: dict = {"sql": sql}
        if params is not None:
            body["params"] = params
        return body

    def query(self, sql: str, params: Optional[list] = None) -> dict:
        return self._map_rows(self._http.request("POST", self._path(), self._body(sql, params)))

    async def aquery(self, sql: str, params: Optional[list] = None) -> dict:
        return self._map_rows(await self._http.request("POST", self._path(), self._body(sql, params)))

    def execute(self, sql: str, params: Optional[list] = None) -> dict:
        return self._http.request("POST", self._path(), self._body(sql, params))

    async def aexecute(self, sql: str, params: Optional[list] = None) -> dict:
        return await self._http.request("POST", self._path(), self._body(sql, params))

    def transaction(self, statements: list[dict]) -> list:
        return self._http.request("POST", f"/v1/databases/{self._cell}/transaction", {"statements": statements})

    async def atransaction(self, statements: list[dict]) -> list:
        return await self._http.request("POST", f"/v1/databases/{self._cell}/transaction", {"statements": statements})


class Kv:
    def __init__(self, http, cell_id: str):
        self._http = http
        self._cell = cell_id

    def _k(self, key: str) -> str:
        return f"/v1/databases/{self._cell}/kv/{key}"

    # ---- 同步 ----
    def get(self, key: str) -> Optional[str]:
        r = self._http.request("GET", self._k(key))
        return r.get("value") if r else None

    def get_json(self, key: str) -> Any:
        import json

        v = self.get(key)
        return json.loads(v) if v is not None else None

    def set(self, key: str, value: str, ttl: Optional[int] = None, condition: Optional[str] = None) -> bool:
        body = {"value": value, "ttl_seconds": ttl, "nx": condition == "nx", "xx": condition == "xx"}
        r = self._http.request("PUT", self._k(key), body)
        return bool(r.get("written"))

    def set_json(self, key: str, value: Any, ttl: Optional[int] = None, condition: Optional[str] = None) -> bool:
        import json

        return self.set(key, json.dumps(value), ttl, condition)

    def delete(self, key: str) -> bool:
        r = self._http.request("DELETE", self._k(key))
        return bool(r.get("deleted"))

    def exists(self, key: str) -> bool:
        r = self._http.request("POST", f"/v1/databases/{self._cell}/kv/ops/exists", {"keys": [key]})
        return bool(r[0])

    def mget(self, keys: list[str]) -> list[Optional[str]]:
        r = self._http.request("POST", f"/v1/databases/{self._cell}/kv/ops/mget", {"keys": keys})
        return list(r.get("values", []))

    def mset(self, entries: dict[str, str]) -> None:
        items = [{"key": k, "value": v} for k, v in entries.items()]
        self._http.request("POST", f"/v1/databases/{self._cell}/kv/ops/mset", {"items": items})

    def ttl(self, key: str) -> dict:
        r = self._http.request("POST", f"/v1/databases/{self._cell}/kv/ops/ttl", {"keys": [key]})
        ttl = r[0]
        if ttl is None:
            return {"state": "missing"}
        if ttl < 0:
            return {"state": "persistent"}
        return {"state": "expires", "seconds": ttl}

    def expire(self, key: str, seconds: int) -> bool:
        r = self._http.request("POST", f"/v1/databases/{self._cell}/kv/ops/expire", {"key": key, "ttl_seconds": seconds})
        return bool(r.get("updated"))

    def persist(self, key: str) -> bool:
        r = self._http.request("POST", f"/v1/databases/{self._cell}/kv/ops/expire", {"key": key, "ttl_seconds": None})
        return bool(r.get("updated"))

    def incr(self, key: str, delta: int = 1) -> int:
        r = self._http.request("POST", f"/v1/databases/{self._cell}/kv/ops/incr", {"key": key, "delta": delta})
        return int(r.get("value"))

    def decr(self, key: str, delta: int = 1) -> int:
        return self.incr(key, -delta)

    # ---- 异步 ----
    async def aget(self, key: str) -> Optional[str]:
        r = await self._http.request("GET", self._k(key))
        return r.get("value") if r else None

    async def aset(self, key: str, value: str, ttl: Optional[int] = None, condition: Optional[str] = None) -> bool:
        body = {"value": value, "ttl_seconds": ttl, "nx": condition == "nx", "xx": condition == "xx"}
        r = await self._http.request("PUT", self._k(key), body)
        return bool(r.get("written"))

    async def adelete(self, key: str) -> bool:
        r = await self._http.request("DELETE", self._k(key))
        return bool(r.get("deleted"))

    async def aincr(self, key: str, delta: int = 1) -> int:
        r = await self._http.request("POST", f"/v1/databases/{self._cell}/kv/ops/incr", {"key": key, "delta": delta})
        return int(r.get("value"))


class Backups:
    def __init__(self, http, cell_id: str):
        self._http = http
        self._cell = cell_id

    def create(self) -> dict:
        return self._http.request("POST", f"/v1/databases/{self._cell}/backup")

    def create_incremental(self) -> dict:
        return self._http.request("POST", f"/v1/databases/{self._cell}/backup/incr")

    def restore(self, version: Optional[str] = None) -> None:
        self._http.request("POST", f"/v1/databases/{self._cell}/restore", {"version": version} if version else {})

    def restore_latest(self) -> None:
        self.restore()


class Replication:
    def __init__(self, http, cell_id: str):
        self._http = http
        self._cell = cell_id

    def get(self) -> dict:
        r = self._http.request("GET", f"/v1/databases/{self._cell}/replication")
        return {
            "enabled": bool(r.get("replica_node")),
            "replica_node": r.get("replica_node"),
        }

    def enable(self, replica_node: str) -> None:
        self._http.request("POST", f"/v1/databases/{self._cell}/replication", {"replica_node": replica_node})

    def disable(self) -> None:
        self._http.request("DELETE", f"/v1/databases/{self._cell}/replication")


class Cell:
    def __init__(self, http, cell_id: str):
        self.id = cell_id
        self.sql = Sql(http, cell_id)
        self.kv = Kv(http, cell_id)
        self.backups = Backups(http, cell_id)
        self.replication = Replication(http, cell_id)
        self._http = http

    def info(self) -> dict:
        all_cells = self._http.request("GET", "/v1/databases?limit=1000")
        for c in all_cells or []:
            if c.get("id") == self.id:
                return c
        from .errors import CellNotFoundError

        raise CellNotFoundError(f"cell not found: {self.id}")

    async def ainfo(self) -> dict:
        all_cells = await self._http.request("GET", "/v1/databases?limit=1000")
        for c in all_cells or []:
            if c.get("id") == self.id:
                return c
        from .errors import CellNotFoundError

        raise CellNotFoundError(f"cell not found: {self.id}")

    def delete(self) -> None:
        self._http.request("DELETE", f"/v1/databases/{self.id}")

    async def adelete(self) -> None:
        await self._http.request("DELETE", f"/v1/databases/{self.id}")


class Cells:
    def __init__(self, http):
        self._http = http

    def create(self, name: Optional[str] = None, region: Optional[str] = None) -> Cell:
        key = f"cell:{name}" if name else None
        r = self._http.request("POST", "/v1/databases", {}, idempotency_key=key)
        return Cell(self._http, r["id"])

    async def acreate(self, name: Optional[str] = None, region: Optional[str] = None) -> Cell:
        key = f"cell:{name}" if name else None
        r = await self._http.request("POST", "/v1/databases", {}, idempotency_key=key)
        return Cell(self._http, r["id"])

    def get(self, cell_id: str) -> Cell:
        cell = Cell(self._http, cell_id)
        cell.info()
        return cell

    async def aget(self, cell_id: str) -> Cell:
        cell = Cell(self._http, cell_id)
        await cell.ainfo()
        return cell

    def list(self, limit: int = 100, cursor: Optional[str] = None) -> dict:
        arr = self._http.request("GET", f"/v1/databases?limit={limit}")
        return {"items": arr or [], "next_cursor": None}

    async def alist(self, limit: int = 100, cursor: Optional[str] = None) -> dict:
        arr = await self._http.request("GET", f"/v1/databases?limit={limit}")
        return {"items": arr or [], "next_cursor": None}

    def delete(self, cell_id: str) -> None:
        self._http.request("DELETE", f"/v1/databases/{cell_id}")

    async def adelete(self, cell_id: str) -> None:
        await self._http.request("DELETE", f"/v1/databases/{cell_id}")

File: sdk/typescript/src/cells.ts
//! Cells / SQL / KV / Backups / Replication(Data Plane + Cell 管理)。

import { Http } from "./http.js";
import type {
  BackupInfo,
  CellInfo,
  CreateCellInput,
  KvGetResult,
  KvSetOptions,
  KvSetResult,
  KvTtl,
  KvValue,
  Page,
  ReplicationStatus,
  SqlExecuteResult,
  SqlParam,
  SqlQueryResult,
  SqlStatement,
} from "./types.js";

/** SQL 命名空间(挂在一个 Cell 下)。 */
export class Sql {
  constructor(
    private http: Http,
    private cellId: string,
  ) {}

  async query<T = Record<string, unknown>>(
    sql: string,
    params?: SqlParam[],
  ): Promise<SqlQueryResult<T>> {
    // 服务端 rows 为位置数组;SDK 按 columns 映射为对象(对齐 SDK_SPEC §8.1)
    const raw = await this.http.request<{ columns: string[]; rows: unknown[][] }>(
      "POST",
      `/v1/databases/${this.cellId}/sql`,
      { sql, params },
    );
    const rows = raw.rows.map((row) =>
      Object.fromEntries(raw.columns.map((c, i) => [c, row[i]])),
    ) as T[];
    return { columns: raw.columns, rows };
  }

  async execute(
    sql: string,
    params?: SqlParam[],
  ): Promise<SqlExecuteResult> {
    return this.http.request<SqlExecuteResult>("POST", `/v1/databases/${this.cellId}/sql`, {
      sql,
      params,
    });
  }

  async transaction(statements: SqlStatement[]): Promise<SqlExecuteResult[]> {
    return this.http.request<SqlExecuteResult[]>(
      "POST",
      `/v1/databases/${this.cellId}/transaction`,
      { statements },
    );
  }
}

/** KV 命名空间(Redis-style,SQLite-backed)。 */
export class Kv {
  constructor(
    private http: Http,
    private cellId: string,
  ) {}

  async get(key: string): Promise<string | null> {
    const r = await this.http.request<KvGetResult>("GET", `/v1/databases/${this.cellId}/kv/${encodeURIComponent(key)}`);
    return r.value;
  }

  async getJson<T = unknown>(key: string): Promise<T | null> {
    const v = await this.get(key);
    return v === null ? null : (JSON.parse(v) as T);
  }

  async set(
    key: string,
    value: KvValue,
    opts: KvSetOptions = {},
  ): Promise<boolean> {
    const r = await this.http.request<KvSetResult>("PUT", `/v1/databases/${this.cellId}/kv/${encodeURIComponent(key)}`, {
      value,
      ttl_seconds: opts.ttl ?? null,
      nx: opts.condition === "nx",
      xx: opts.condition === "xx",
    });
    return r.written;
  }

  async setJson(key: string, value: unknown, opts: KvSetOptions = {}): Promise<boolean> {
    return this.set(key, JSON.stringify(value), opts);
  }

  async delete(key: string): Promise<boolean> {
    const r = await this.http.request<{ deleted: boolean }>("DELETE", `/v1/databases/${this.cellId}/kv/${encodeURIComponent(key)}`);
    return r.deleted;
  }

  async exists(key: string): Promise<boolean> {
    const r = await this.http.request<boolean[]>("POST", `/v1/databases/${this.cellId}/kv/ops/exists`, { keys: [key] });
    return r[0];
  }

  async mget(keys: string[]): Promise<Array<string | null>> {
    const r = await this.http.request<{ values: Array<string | null> }>("POST", `/v1/databases/${this.cellId}/kv/ops/mget`, { keys });
    return r.values;
  }

  async mset(entries: Record<string, string>): Promise<void> {
    const items = Object.entries(entries).map(([key, value]) => ({ key, value }));
    await this.http.request("POST", `/v1/databases/${this.cellId}/kv/ops/mset`, { items });
  }

  async ttl(key: string): Promise<KvTtl> {
    // 服务端 ttl 为批量端点:{keys:[...]} → Vec<Option<i64>>
    const r = await this.http.request<Array<number | null>>("POST", `/v1/databases/${this.cellId}/kv/ops/ttl`, { keys: [key] });
    const ttl = r[0];
    // 服务端语义:null = key 不存在;-1 = persistent;>=0 = 剩余秒
    if (ttl === null) return { state: "missing" };
    if (ttl < 0) return { state: "persistent" };
    return { state: "expires", seconds: ttl };
  }

  async expire(key: string, seconds: number): Promise<boolean> {
    const r = await this.http.request<{ updated: boolean }>("POST", `/v1/databases/${this.cellId}/kv/ops/expire`, { key, ttl_seconds: seconds });
    return r.updated;
  }

  async persist(key: string): Promise<boolean> {
    // ttl_seconds: null = 移除过期(服务端 Option<u64> None = persist)
    const r = await this.http.request<{ updated: boolean }>("POST", `/v1/databases/${this.cellId}/kv/ops/expire`, { key, ttl_seconds: null });
    return r.updated;
  }

  async incr(key: string, delta = 1): Promise<number> {
    const r = await this.http.request<{ value: number }>("POST", `/v1/databases/${this.cellId}/kv/ops/incr`, { key, delta });
    return r.value;
  }

  async decr(key: string, delta = 1): Promise<number> {
    return this.incr(key, -delta);
  }
}

/** 备份/恢复(对象存储)。 */
export class Backups {
  constructor(
    private http: Http,
    private cellId: string,
  ) {}

  async create(): Promise<BackupInfo> {
    return this.http.request<BackupInfo>("POST", `/v1/databases/${this.cellId}/backup`);
  }

  async createIncremental(): Promise<BackupInfo> {
    return this.http.request<BackupInfo>("POST", `/v1/databases/${this.cellId}/backup/incr`);
  }

  async restore(version?: string): Promise<void> {
    await this.http.request("POST", `/v1/databases/${this.cellId}/restore`, version ? { version } : {});
  }

  async restoreLatest(): Promise<void> {
    await this.restore();
  }
}

/** 复制状态(单 replica)。 */
export class Replication {
  constructor(
    private http: Http,
    private cellId: string,
  ) {}

  async get(): Promise<ReplicationStatus> {
    const r = await this.http.request<{ replica_node: string | null } & Record<string, unknown>>("GET", `/v1/databases/${this.cellId}/replication`);
    return {
      enabled: Boolean(r.replica_node),
      replica_node: r.replica_node ?? undefined,
      ...("storage_node_id" in r ? { primary_node: String(r.storage_node_id) } : {}),
      ...("generation" in r ? { generation: Number(r.generation) } : {}),
    };
  }

  async enable(replicaNode: string): Promise<void> {
    await this.http.request("POST", `/v1/databases/${this.cellId}/replication`, { replica_node: replicaNode });
  }

  async disable(): Promise<void> {
    await this.http.request("DELETE", `/v1/databases/${this.cellId}/replication`);
  }
}

/** 一个 Cell 的本地句柄(可延迟绑定,首次调用才发请求)。 */
export class Cell {
  readonly sql: Sql;
  readonly kv: Kv;
  readonly backups: Backups;
  readonly replication: Replication;

  constructor(
    private http: Http,
    readonly id: string,
  ) {
    this.sql = new Sql(http, id);
    this.kv = new Kv(http, id);
    this.backups = new Backups(http, id);
    this.replication = new Replication(http, id);
  }

  async info(): Promise<CellInfo> {
    // 通过列表取回该 Cell 信息(无单点 GET 端点时使用)
    const all = await this.http.request<CellInfo[]>("GET", `/v1/databases?limit=1000`);
    const found = all.find((c) => c.id === this.id);
    if (!found) {
      throw new (await import("./errors.js")).CellNotFoundError(`cell not found: ${this.id}`);
    }
    return found;
  }

  async delete(): Promise<void> {
    await this.http.request("DELETE", `/v1/databases/${this.id}`);
  }
}

/** Cells 资源(租户级)。 */
export class Cells {
  constructor(private http: Http) {}

  async create(input?: CreateCellInput): Promise<Cell> {
    const r = await this.http.request<{ id: string }>(
      "POST",
      "/v1/databases",
      {},
      { idempotencyKey: input?.name ? `cell:${input.name}` : undefined },
    );
    return new Cell(this.http, r.id);
  }

  async get(id: string): Promise<Cell> {
    const cell = new Cell(this.http, id);
    await cell.info();
    return cell;
  }

  async list(limit = 100, cursor?: string): Promise<Page<CellInfo>> {
    // 服务端 list 返回数组(非游标分页);包装为 Page 以对齐 SDK_SPEC
    void cursor;
    const sep = "?";
    const arr = await this.http.request<CellInfo[]>("GET", `/v1/databases${sep}limit=${limit}`);
    return { items: arr, nextCursor: null };
  }

  async delete(id: string): Promise<void> {
    await this.http.request("DELETE", `/v1/databases/${id}`);
  }
}
